packages/@cruise-automation/worker-rpc/src/index.js

```diff
@@ -0,0 +1,138 @@
+// @flow
+//
+//  Copyright (c) 2018-present, GM Cruise LLC
+//
+//  This source code is licensed under the Apache License, Version 2.0,
+//  found in the LICENSE file in the root directory of this source tree.
+//  You may not use this file except in compliance with the License.
+
+// this type mirrors the MessageChannel api which is available on
+// instances of web-workers as well as avaiable on 'global' within a worker
+
+export type Channel = {
+  postMessage: (data: any, transfer?: any[]) => void,
+  onmessage?: (ev: MessageEvent) => void,
+};
+
+const RESPONSE = "$$RESPONSE";
+const ERROR = "$$ERROR";
+
+// helper function to create linked channels for testing
+export function createLinkedChannels(): { local: Channel, remote: Channel } {
+  const local: Channel = {
+    postMessage(data: any, transfer?: Array<ArrayBuffer>) {
+      const ev = new MessageEvent("message", { data });
+      // eslint-disable-next-line no-use-before-define
+      if (remote.onmessage) {
+        remote.onmessage(ev); // eslint-disable-line no-use-before-define
+      }
+    },
+  };
+
+  const remote: Channel = {
+    postMessage(data: any, transfer?: Array<ArrayBuffer>) {
+      const ev = new MessageEvent("message", { data });
+      if (local.onmessage) {
+        local.onmessage(ev);
+      }
+    },
+  };
+  return { local, remote };
+}
+
+// This class allows you to hook up bi-directional async calls across web-worker
+// boundaries where a single call to or from a worker can 'wait' on the response.
+// Errors in receivers are propigated back to the caller as a rejection.
+// It also supports returning transferables over the web-worker postMessage api,
+// which was the main shortcomming with the worker-rpc npm module.
+// To attach rpc to an instance of a worker in the main thread:
+//   const rpc = new Rpc(workerInstace);
+// To attach rpc within an a web worker:
+//   const rpc = new Rpc(global);
+// Check out the tests for more examples.
+export default class Rpc {
+  static transferables = "$$TRANSFERABLES";
```

Comment:
is it possible to use `Symbol` for this, or does that break with MessagePort?

Reply:
Good call - should be fine since we don't actually _send_ this key across the message port. As far as it working across message ports I'd guess it _doesn't_ since it creates a unique value for every constructor invocation and there is a separate instance of `rpc` on both sides of the boundary.

Reply:
ehh...flow doesn't like using a symbol for a computed object property.  While it's technically valid JS I'd rather avoid using it since we'd have to attach `// $FlowFixMe` everywhere we used this.  I don't see it being a huge benefit versus the cost, so I'll just leave it as a string for now.  

https://github.com/facebook/flow/issues/3258